src/journal/JournalMetadata.cc

```diff
@@ -748,7 +751,13 @@ void JournalMetadata::handle_refresh_complete(C_Refresh *refresh, int r) {
 
     Client client(m_client_id, bufferlist());
     RegisteredClients::iterator it = refresh->registered_clients.find(client);
-    if (it != refresh->registered_clients.end()) {
+    if (it == refresh->registered_clients.end()) {
+      lderr(m_cct) << "failed to locate client: " << m_client_id << dendl;
+      r = -ENOENT;
+    } else if (it->state == cls::journal::CLIENT_STATE_DISCONNECTED) {
+      lderr(m_cct) << "client flagged disconnected: " << m_client_id << dendl;
+      r = -ENOTCONN;
```

Comment:
Even if the client isn't connected, I think you would still want to initialize the journaler. rbd-mirror, during bootstrap can check its connected and use a helper method to re-connect if disconnected and within ImageReplayer it should have a listener for journal updates to check if it has been disconnected while replaying.